Rewrite `RebalanceIntegrationTests`; filter out buffered records from recently revoked partitions

Aaronontheweb opened the pull request:
## Changes

Working on simplifying the `RebalanceSpec` so it's more deterministic and doesn't take as long to run either.

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).

Aaronontheweb committed:
WIP - designing custom partitioning strategy

Aaronontheweb commented:
FYI it's possible with the .NET Kafka client to customize partition assignments

https://github.com/confluentinc/confluent-kafka-dotnet/blob/7ae5242ce953c2929285f440878753236dc99bc1/src/Confluent.Kafka/ConsumerBuilder.cs#L330-L341

Arkatufus commented:
Custom partition assignment (assignor) is not possible in the C# driver: https://github.com/confluentinc/confluent-kafka-dotnet/issues/1589#issuecomment-823948005

Aaronontheweb committed:
BUGFIX: make `TopicSubscription` actually immutable

Aaronontheweb committed:
WIP trying to refactor specs

Aaronontheweb committed:
stash

Aaronontheweb committed:
fixed several stage definitions

Aaronontheweb committed:
fixed base single source logic

Aaronontheweb committed:
fixed compilation errors

Aaronontheweb committed:
integrated updated `IPartitionEventHandler` into `ISubscriptionRequest` messages

Aaronontheweb committed:
Added rebalace notification actor support to `AsyncCallbacks`

Aaronontheweb committed:
no need for separate `OnPartitionLost` event handler

we can just call `OnPartitionsRevoked` when the `OnPartitionsLost` event is raised

Aaronontheweb committed:
cleaned up spec and event types

Aaronontheweb committed:
Added `verbose-logging` setting

Aaronontheweb committed:
we do not need a `ConcurrentQueue<T>` internally

Aaronontheweb committed:
ensured that records from revoked partitions get filtered from the stream upon rebalance

Aaronontheweb committed:
have a failing test

filtering has not been implemented yet in sub-sources

Aaronontheweb committed:
disable filtering on sub-sources

Aaronontheweb commented:
This is ready for review - any of the JVM sources I referenced were all from https://github.com/akka/alpakka-kafka/releases/tag/v3.0.1 - which is before the license changed from Apache 2.0.

Aaronontheweb commented on src/Akka.Streams.Kafka.Tests/Integration/RebalanceIntegrationTests.cs:
```diff
@@ -22,110 +28,211 @@ public RebalanceIntegrationTests(ITestOutputHelper output, KafkaFixture fixture)
     /// </summary>
     [Fact]
-    public async Task ShouldReBalanceWithoutArgumentExceptions()
+    public async Task FetchedRecords_must_be_removed_from_source_stage_buffer_when_partition_is_removed()
```
Spec has been rewritten to more or less match the rebalance spec - and in this case it's checking to ensure that the individual stream stages do their level best to filter out messages from revoked partitions before emitting them downstream. This is all with the intention of eliminating duplicates, as those messages will be picked up by someone else.

Aaronontheweb commented on src/Akka.Streams.Kafka.Tests/Integration/RebalanceIntegrationTests.cs:
```diff
@@ -22,110 +28,211 @@ public RebalanceIntegrationTests(ITestOutputHelper output, KafkaFixture fixture)
+        await probe1.CancelAsync();
+        await probe2.CancelAsync();
+
+        // TODO: bug - if we've cancelled the stream, the controls should complete automatically
```
Found a bug here - will have to fix later

Aaronontheweb commented on src/Akka.Streams.Kafka.Tests/Integration/RebalanceIntegrationTests.cs:
```diff
@@ -22,110 +28,211 @@ public RebalanceIntegrationTests(ITestOutputHelper output, KafkaFixture fixture)
+        // Assert.True(control2.IsShutdown.IsCompleted);
+    }
+
+    [Fact(Skip = "Filtering is not enabled on sub-sources just yet")]
```
Not working yet because honestly the sub-sources are a mess and need some major work before I can even fix them correctly.

Aaronontheweb commented on src/Akka.Streams.Kafka/Helpers/PartitionEventHandler.cs:
```diff
@@ -41,15 +42,21 @@ public interface IPartitionEventHandler
         /// Dummy handler which does nothing. Also <see cref="IPartitionEventHandler"/>
         /// </summary>
-        internal class Empty : IPartitionEventHandler
+        internal sealed class Empty : IPartitionEventHandler
```
Made this into a singleton

Aaronontheweb commented on src/Akka.Streams.Kafka/Helpers/PartitionEventHandler.cs:
```diff
@@ -69,54 +76,69 @@ public void OnAssign(IImmutableSet<TopicPartition> assignedTopicPartitions, IRes
+            
             private readonly Action<IImmutableSet<TopicPartition>> _partitionAssignedCallback;
             private readonly Action<IImmutableSet<TopicPartitionOffset>> _partitionRevokedCallback;
-            private readonly Action<IImmutableSet<TopicPartitionOffset>> _partitionLostCallback;
```
We do not need a separate "partitions lost" handler - just calling the "partitions revoked" handler when a partitions lost event is raised is sufficient.

Aaronontheweb commented on src/Akka.Streams.Kafka/Helpers/PartitionEventHandler.cs:
```diff
@@ -69,54 +76,69 @@ public void OnAssign(IImmutableSet<TopicPartition> assignedTopicPartitions, IRes
+            
             public void OnRevoke(IImmutableSet<TopicPartitionOffset> revokedTopicPartitions, IRestrictedConsumer consumer)
             {
+                _subscription.RebalanceListener
```
Added support for the `Option<IActorRef>` rebalance listener - which is very useful for testing.

Aaronontheweb commented on src/Akka.Streams.Kafka/Settings/ConsumerSettings.cs:
```diff
@@ -69,8 +70,12 @@ public static ConsumerSettings<TKey, TValue> Create(Akka.Configuration.Config co
+                    config.GetConfig(ConnectionCheckerSettings.ConfigPath)),
+                consumerFactory: null)
+            {
+                VerboseLogging = config.GetBoolean("verbose-logging", false)
```
Added a new setting - `VerboseLogging`. This is designed to eliminate `KafkaConsumerActor` spam.

Aaronontheweb commented on src/Akka.Streams.Kafka/Settings/Subscriptions.cs:
```diff
@@ -24,20 +26,36 @@ public interface IManualSubscription : ISubscription { }
+        /// <summary>
+        /// Optional actor that receives rebalance events as messages.
+        /// </summary>
+        Option<IActorRef> RebalanceListener { get; }
```
The `RebalanceListener` is part of the `IAutoSubscription` and this gets called from inside the `AsyncCallbacks: IPartitionEventHandler`

Aaronontheweb commented on src/Akka.Streams.Kafka/Settings/Subscriptions.cs:
```diff
@@ -24,20 +26,36 @@ public interface IManualSubscription : ISubscription { }
+    
+    public interface IConsumerRebalanceEvent;
+    
+    public sealed record TopicPartitionsAssigned(ISubscription Subscription, IImmutableSet<TopicPartition> Partitions) : IConsumerRebalanceEvent;
```
consumer rebalance events

Aaronontheweb commented on src/Akka.Streams.Kafka/Settings/Subscriptions.cs:
```diff
@@ -24,20 +26,36 @@ public interface IManualSubscription : ISubscription { }
+    /// A subscription to a set of 1 or more topics.
     /// </summary>
-    internal sealed class TopicSubscription : IAutoSubscription
+    internal sealed record TopicSubscription : IAutoSubscription
```
Fixed some mutability bugs here. Internal API so no breaking change.

Aaronontheweb commented on src/Akka.Streams.Kafka/Settings/Subscriptions.cs:
```diff
@@ -80,7 +103,7 @@ public IAutoSubscription WithPartitionEventsHandler(IPartitionEventHandler parti
     /// Allows subscription to multiple topics, matching given regex pattern
     /// </remarks>
-    internal sealed class TopicSubscriptionPattern : IAutoSubscription
+    internal sealed record TopicSubscriptionPattern : IAutoSubscription
```
Fixed some mutability bugs here. Internal API so no breaking change.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/BaseSingleSourceLogic.cs:
```diff
@@ -26,10 +26,30 @@ internal abstract class BaseSingleSourceLogic<K, V, TMessage> : GraphStageLogic
 
-        private readonly ConcurrentQueue<ConsumeResult<K, V>> _buffer = new ConcurrentQueue<ConsumeResult<K, V>>();
-        protected IImmutableSet<TopicPartition> TopicPartitions { get; set; } = ImmutableHashSet.Create<TopicPartition>();
+        public Action<IImmutableSet<TopicPartitionOffset>> FilterRevokedPartitionAsyncCallback =>
```
Filtering callback for after partitions are revoked.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/BaseSingleSourceLogic.cs:
```diff
@@ -26,10 +26,30 @@ internal abstract class BaseSingleSourceLogic<K, V, TMessage> : GraphStageLogic
+        protected ISubscription Subscription { get; }
         protected Decider Decider { get; }
 
-        private readonly ConcurrentQueue<ConsumeResult<K, V>> _buffer = new ConcurrentQueue<ConsumeResult<K, V>>();
```
This never needed to be a thread-safe collection.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/BaseSingleSourceLogic.cs:
```diff
@@ -26,10 +26,30 @@ internal abstract class BaseSingleSourceLogic<K, V, TMessage> : GraphStageLogic
         private readonly IMessageBuilder<K, V, TMessage> _messageBuilder;
         private int _requestId = 0;
         private bool _requested = false;
+        protected ISubscription Subscription { get; }
```
We store the `ISubscription` as a protected property now because it gets used in the `ConfigureSubscription` method.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/BaseSingleSourceLogic.cs:
```diff
@@ -85,6 +108,47 @@ public override void PostStop()
         /// </summary>
         protected abstract void ConfigureSubscription();
 
+        protected void ConfigureSubscription(Action<IImmutableSet<TopicPartition>> partitionsAssignedCb,
```
overload of `ConfigureSubscription` that automatically sets up all `ISubscriptionRequest`s with the `KafkaConsumerActor`.

Worth noting that callback handlers et al are all now set through the `Subscribe` / etc messages and are not passed into the `KafkaConsumerActor`'s CTOR anymore.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/BaseSingleSourceLogic.cs:
```diff
@@ -85,6 +108,47 @@ public override void PostStop()
+
+            IPartitionEventHandler CreateRebalanceListener(IAutoSubscription subscription)
+            {
+                return new PartitionEventHandlers.Chain(subscription.PartitionEventsHandler.GetOrElse(PartitionEventHandlers.Empty.Instance), new PartitionEventHandlers.AsyncCallbacks(subscription, SourceActor.Ref, partitionsAssignedCb,
```
Ensure that our async callbacks always get invoked.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs:
```diff
@@ -22,67 +22,53 @@ internal class SingleSourceStageLogic<K, V, TMessage> : BaseSingleSourceLogic<K,
 
-        /// <inheritdoc />
-        protected override void ConfigureSubscription()
+        protected override object LogSource
```
Customized the `LogSource` to make it more obvious who is doing the logging here.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs:
```diff
@@ -22,67 +22,53 @@ internal class SingleSourceStageLogic<K, V, TMessage> : BaseSingleSourceLogic<K,
         /// <inheritdoc />
         protected override IActorRef CreateConsumerActor()
         {
-            var partitionsAssignedHandler = GetAsyncCallback<IEnumerable<TopicPartition>>(PartitionsAssigned);
```
All of the event handler configuration code got moved to `ConfigureSubscription`, which happens after the `KafkaConsumerActor` is created.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs:
```diff
@@ -22,67 +22,53 @@ internal class SingleSourceStageLogic<K, V, TMessage> : BaseSingleSourceLogic<K,
             
             var extendedActorSystem = actorMaterializer.System.AsInstanceOf<ExtendedActorSystem>();
-            var actor = extendedActorSystem.SystemActorOf(KafkaConsumerActorMetadata.GetProps(SourceActor.Ref, _settings, Decider, eventHandler, statisticsHandler),
+            var actor = extendedActorSystem.SystemActorOf(KafkaConsumerActorMetadata.GetProps(SourceActor.Ref, _settings, Decider, statisticsHandler),
```
Stats handler still gets passed in via CTOR since it's not really re-configurable the same way the `IPartitionHandler`s are.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs:
```diff
@@ -134,23 +112,52 @@ protected virtual void StopConsumerActor()
         }
 
-        private void PartitionsAssigned(IEnumerable<TopicPartition> partitions)
+        private class FlushMessagesOfRevokedPartitionsHandler : IPartitionEventHandler
```
Custom `IPartitionEventHandler` for filtering out messages from recently revoked partitions.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs:
```diff
@@ -134,23 +112,52 @@ protected virtual void StopConsumerActor()
+            public void OnRevoke(IImmutableSet<TopicPartitionOffset> revokedTopicPartitions,
+                IRestrictedConsumer consumer)
+            {
+                _lastRevoked = revokedTopicPartitions;
```
Memorize which partitions are revoked

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs:
```diff
@@ -134,23 +112,52 @@ protected virtual void StopConsumerActor()
+            public void OnAssign(IImmutableSet<TopicPartition> assignedTopicPartitions, IRestrictedConsumer consumer)
+            {
+                // remove all of our previous revoked partitions that are not in the new assignment
+                _stageLogic.FilterRevokedPartitionAsyncCallback(_lastRevoked
```
Revoke messages from all partitions that were not assigned to us - when the Kafka broker rebalances partitions all partitions get revoked from every body, but one of the original partitions we owned might get assigned back to us again, so we don't want to discard messages until we're sure about which partitions we're keeping.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs:
```diff
@@ -134,23 +112,52 @@ protected virtual void StopConsumerActor()
+
+            public void OnLost(IImmutableSet<TopicPartitionOffset> revokedTopicPartitions, IRestrictedConsumer consumer)
+            {
+                _stageLogic.FilterRevokedPartitionAsyncCallback(revokedTopicPartitions);
```
If we've lost a partition, we can filter those messages immediately.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs:
```diff
@@ -134,23 +112,52 @@ protected virtual void StopConsumerActor()
-        
-        private void PartitionsRevoked(IEnumerable<TopicPartitionOffset> partitions)
+
+        protected override IPartitionEventHandler AddToPartitionAssignmentHandler(IPartitionEventHandler handler) => 
```
Chain the filtering system together with any other handlers that were already defined.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs:
```diff
@@ -134,23 +112,52 @@ protected virtual void StopConsumerActor()
+            RequestMessages();
         }
         
-        private void PartitionsLost(IEnumerable<TopicPartitionOffset> partitions)
```
Remove the separate `PartitionsLost` method and just consolidate it with `PartitionsRevoked` - it's not a meaningful difference to us.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SubSourceLogic.cs:
I have more work to do around the `SubSourceLogic` - but since we don't have Scala traits the best I can do is have it copy and paste a lot of the code from the `SingleSourceStageLogic` / `BaseSingleSourceStageLogic`.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Abstract/TransactionalSourceLogic.cs:
```diff
@@ -139,13 +139,28 @@ protected override void StopConsumerActor()
-                partitionAssignedCallback: _ => { },
-                partitionRevokedCallback: OnRevoke,
-                partitionLostCallback: OnRevoke);
+            var blockingRevokedCall = new TransactionalAssignmentHandler(OnRevoke);
```
Make sure we're still taking care of the transactional assignment handler

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:
```diff
@@ -39,7 +37,7 @@ internal class KafkaConsumerActor<K, V> : ActorBase, ILogReceive, IWithTimers
         /// Stores delegates for external handling of partition events
         /// </summary>
-        private readonly IPartitionEventHandler _partitionEventHandler;
+        private IPartitionEventHandler _partitionEventHandler;
```
The `_partitionEventHandler` will now be assigned via message-handling rather than through the CTOR

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:
```diff
@@ -185,27 +183,6 @@ protected override bool Receive(object message)
         {
             switch (message)
             {
-                case KafkaConsumerActorMetadata.Internal.Assign assign:
```
These are really `ISubscriptionRequest` messages so they've been moved to the `HandleSubscriptions` method.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:
```diff
@@ -371,14 +348,45 @@ private void HandleSubscription(KafkaConsumerActorMetadata.Internal.ISubscriptio
+                    case KafkaConsumerActorMetadata.Internal.Subscribe subscribe:
+                    {
+                        _consumer.Subscribe(subscribe.Topics);
+                        _partitionEventHandler = subscribe.RebalanceHandler;
```
Handlers are set here.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:
```diff
@@ -371,14 +348,45 @@ private void HandleSubscription(KafkaConsumerActorMetadata.Internal.ISubscriptio
+                        var topicPartitions = assignWithOffset.TopicPartitionOffsets.Select(o => o.TopicPartition).ToImmutableHashSet();
+                        CheckOverlappingRequests("AssignWithOffset", Sender, topicPartitions);
+                        
+                        // TODO: dear lord this is wrong, WE SHOULD NOT BE SETTING OFFSETS TO ZERO HERE
```
TL;DR; not sure how bad of a bug this actually is but it is giving off code smell.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs:
```diff
@@ -469,7 +477,7 @@ private void Poll()
             if (partitionsToFetch.IsEmpty || _requests.IsEmpty())
             {
-                if(_log.IsDebugEnabled)
+                if(_settings.VerboseLogging)
```
Verbose logging setting in-use to reduce log spam.

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActorMetadata.cs:
```diff
@@ -23,260 +23,89 @@ public static class KafkaConsumerActorMetadata
-            Props.Create(() => new KafkaConsumerActor<K, V>(owner, settings, decider, handler, statisticsHandler)).WithDispatcher(settings.DispatcherId);
+            GetProps(null, settings, decider, null);
+   
+        internal static Props GetProps<K, V>(IActorRef? owner, ConsumerSettings<K, V> settings, Decider decider, IStatisticsHandler? statisticsHandler) =>
```
Simplified `KafkaConsumerActor` `Props`

Aaronontheweb commented on src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActorMetadata.cs:
```diff
@@ -23,260 +23,89 @@ public static class KafkaConsumerActorMetadata
         /// </summary>
         [InternalApi]
-        public class Internal
+        public static class Internal
```
All of these messages should really have `INoSerializationVerificationNeeded` - so I added that and transformed them into `record` types as well.

Aaronontheweb commented on src/Akka.Streams.Kafka/reference.conf:
```diff
@@ -54,6 +54,10 @@ akka.kafka.consumer {
   
+  # When enabled, the client will emit very detailed trace information
+  # Helpful for debugging, but do not recommend running in production
+  verbose-logging = false
```
verbose-logging HOCON setting.

Aaronontheweb reviewed (commented):
Detailed my changes

Aaronontheweb commented:
Test that is failing has been a consistently racy spec - and I think I'll probably get to fix it when I work on partitioned sources next
